src/autocomplete.js

```diff
@@ -187,11 +193,82 @@ class Autocomplete extends Component {
   }
 
   selectHighlightedItem = () => {
-    return this.selectItemAtIndex(this.state.highlightedIndex)
+    return this.selectItemAtIndex(this.getState().highlightedIndex)
+  }
+
+  // any piece of our state can live in two places:
+  // 1. Uncontrolled: it's internal (this.state)
+  //    We will call this.setState to update that state
+  // 2. Controlled: it's external (this.props)
+  //    We will call this.props.onChange to update that state
+  //
+  // In addition, we'll always call this.props.onChange if the
+  // selectedValue is changed because that's important whether
+  // that property is controlled or not.
+  internalSetState(stateToSet, cb) {
+    const onChangeArg = {}
+    let onStateChangeArg
+    return this.setState(
+      state => {
+        state = this.mergeState(state)
+        onStateChangeArg =
+          typeof stateToSet === 'function' ? stateToSet(state) : stateToSet
+        const nextState = {}
+        // we need to call on change if the outside world is controlling any of our state
+        // and we're trying to update that state. OR if the selection has changed and we're
+        // trying to update the selection
+        if (onStateChangeArg.hasOwnProperty('selectedValue')) {
+          onChangeArg.selectedValue = onStateChangeArg.selectedValue
+          onChangeArg.previousValue = state.selectedValue
+        }
+        Object.keys(onStateChangeArg).forEach(key => {
+          // if it's coming from props, then we don't want to set it internally
+          if (!this.props.hasOwnProperty(key)) {
+            nextState[key] = onStateChangeArg[key]
+          }
+        })
+        return nextState
+      },
+      () => {
+        // call the provided callback if it's a callback
+        cbToCb(cb)()
+        // if the selectedValue changed
+        // then let's call onChange!
+        if (Object.keys(onChangeArg).length) {
+          this.props.onChange(onChangeArg)
+        }
+        // We call this function whether we're controlled or not
+        // It's mostly useful if we're controlled, but it can
+        // definitely be useful for folks to know when something
+        // happens internally.
+        this.props.onStateChange(onStateChangeArg)
```

Comment:
Hmmm, if I'm following your argument correctly, you're trying to avoid passing anything to the parent that was never actually the state of this component.

However we're doing exactly that: passing properties to the parent that maybe never actually were part of the state.

Also: if this is a totally controlled component, it's possible that we never actually change the state, right?

Currently I'm not sure if state and props give us all the tools necessary to avoid such interesting effects and we'd need to keep the "state" in a separate property of the component. Does that make sense?